src/Microsoft.AspNetCore.Blazor/Components/ComponentResolver.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Blazor.Components
+{
+    /// <summary>
+    /// Resolves components for an application.
+    /// </summary>
+    internal class ComponentResolver
+    {
+        /// <summary>
+        /// Lists all the types 
+        /// </summary>
+        /// <param name="appAssembly"></param>
+        /// <returns></returns>
+        public static IEnumerable<Type> ResolveComponents(Assembly appAssembly)
+        {
+            var blazorAssembly = typeof(IComponent).Assembly;
+
+            return EnumerateAssemblies(appAssembly.GetName(), blazorAssembly, new HashSet<Assembly>(new AssemblyComparer()))
+                .SelectMany(a => a.ExportedTypes)
+                .Where(t => typeof(IComponent).IsAssignableFrom(t));
+        }
+
+        private static IEnumerable<Assembly> EnumerateAssemblies(
+            AssemblyName assemblyName,
+            Assembly blazorAssembly,
+            HashSet<Assembly> visited)
+        {
+            var assembly = Assembly.Load(assemblyName);
+            if (visited.Contains(assembly))
+            {
+                // Avoid traversing visited assemblies.
+                yield break;
+            }
+            visited.Add(assembly);
+            var references = assembly.GetReferencedAssemblies();
+            if (!references.Any(r => string.Equals(r.FullName, blazorAssembly.FullName, StringComparison.Ordinal)))
```

Comment:
I'm curious about this logic. What if there was the following dependency graph?

  MyApp -> SuperVendorComponents -> SuperVendorComponentsImpl -> Microsoft.AspNetCore.Blazor

Wouldn't this fail to fetch the components from `SuperVendorComponents` if it only references `SuperVendorComponentsImpl` and doesn't have a direct reference to `Microsoft.AspNetCore.Blazor`?

I'm guessing that `GetReferencedAssemblies` doesn't include transitive references, because if it did then you wouldn't need line 53.

Reply:
On further consideration, I've realised it's probably better if we *don't* scan any referenced assemblies. We should only include components in `AppAssembly`. This is so that, in the future, we can add a feature where you explicitly opt in to mounting other assemblies at specified points in the URI space. For example, you might choose to attach "MyCompany.MyProduct.Administration" with the URI prefix `/admin`.

So for now I think we can avoid the complexity here entirely :)

Reply:
How would you split the implementation between supervendorcomponents and supervendorcomponentsimpl. Isn't supervendorcomponentsimpl the only thing that matters?

That said, we do a similar thing on MVC, we only take into account libraries that reference at least one of the MVC libraries directly.

Reply:
Let's not worry about those details. For now we just don't consider any assembly other than `AppAssembly`.

Reply:
**Update**: Changed my mind. I think this is fine as-is. Let me know if you want to discuss reasoning.